File: src/main/java/no/nav/pensjon/innsyn/sql/DbEntityGetter.java
package no.nav.pensjon.innsyn.sql;

import no.nav.pensjon.innsyn.source.EntityGetter;
import no.nav.pensjon.innsyn.source.SourceException;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

public abstract class DbEntityGetter<T> implements EntityGetter<T> {

    private final Connection connection;

    protected DbEntityGetter(Connection connection) {
        this.connection = connection;
    }

    public abstract String getSql();

    public abstract Function<ResultSet, T> getMap();

    @Override
    public List<T> getEntities() {
        try (PreparedStatement statement = connection.prepareStatement(getSql());
             ResultSet resultSet = statement.executeQuery()) {
            return extractEntitiesFrom(resultSet);
        } catch (SQLException e) {
            throw new SourceException("Fetching entities from DB failed: " + e.getMessage(), e);
        }
    }

    private List<T> extractEntitiesFrom(ResultSet source) throws SQLException {
        List<T> entities = new ArrayList<>();

        while (source.next()) {
            entities.add(getMap().apply(source));
        }

        return entities;
    }
}

File: src/main/kotlin/no/nav/pensjon/innsyn/sql/DbEntityGetter.kt
package no.nav.pensjon.innsyn.sql

import no.nav.pensjon.innsyn.source.EntityGetter
import no.nav.pensjon.innsyn.source.SourceException
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.SQLException
import java.util.*

abstract class DbEntityGetter<T> protected constructor(private val connection: Connection) : EntityGetter<T> {
    abstract val sql: String
    abstract fun map(source: ResultSet): T
    override val entities: List<T>
        get() =
            try {
                connection.prepareStatement(sql)
                        .use(PreparedStatement::executeQuery)
                        .use(this::extractEntitiesFrom)
            } catch (e: SQLException) {
                throw SourceException("Fetching entities from DB failed: ${e.message}", e)
            }

    @Throws(SQLException::class)
    private fun extractEntitiesFrom(source: ResultSet) =
        ArrayList<T>().apply{
            while (source.next()) add(map(source))
        }

}